8321931: memory_swap_current_in_bytes reports 0 as "unlimited"

gerard-ziemski opened the pull request.

gerard-ziemski commented:
@iklam @poonamparhar @jerboaa can you please review this fix? It's pretty simple bug fix I think.

tstuefe commented on src/hotspot/os/linux/osContainer_linux.cpp:
```diff
@@ -137,9 +137,9 @@ jlong OSContainer::pids_current() {
+void OSContainer::print_container_helper(outputStream* st, jlong j, const char* metrics, boolean limit) {
   st->print("%s: ", metrics);
-  if (j > 0) {
+  if ((j > 0) || ((j == 0) && (limit == false))) {
```
Clearer would be a reversal like this:
```
if (j == 0 && limit) {
  "unlimited"
} else {
  print
}
```

gerard-ziemski replied:
I think I decided to have it this way to draw the attention to the fact that the value `0` for a `non-limit` kind should be treated as numerical value and not a special case.

This provides, in my opinion, a self-documenting code without the need for an explicit comment. That was my rationale at least, but if that doesn't come across, then we can simplify and add an actual comment?

tstuefe replied:
Up to you. Both are okay, this was just a small nit :)

tstuefe commented on src/hotspot/os/linux/osContainer_linux.hpp:
```diff
@@ -45,7 +45,7 @@ class OSContainer: AllStatic {
   static void init();
   static void print_version_specific_info(outputStream* st);
-  static void print_container_helper(outputStream* st, jlong j, const char* metrics);
+  static void print_container_helper(outputStream* st, jlong j, const char* metrics, boolean limit = false);
```
Can we name this parameter better, or add a comment at least?

gerard-ziemski replied:
Would `isLimit` be any better?

tstuefe replied:
Yes

tstuefe reviewed (commented):
Are we sure 0 means unlimited? I looked at https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt
and did not find any remarks.

https://unix.stackexchange.com/questions/420906/what-is-the-value-for-the-cgroups-limit-in-bytes-if-the-memory-is-not-restricte seems to indicate the value is some very high number. Or is this handled by the JVMs cgroup layer?

jerboaa commented:
I'll have a look tomorrow. In general unlimited means `-1` not `0`. cg v1 doesn't explicitly set it to a specified value if it is unlimited. It's a very large number. For cg v2 it's usually the string `max` as specified by the kernel docs of interface files.

gerard-ziemski commented:
> Are we sure 0 means unlimited? I looked at https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt and did not find any remarks.
> 
> https://unix.stackexchange.com/questions/420906/what-is-the-value-for-the-cgroups-limit-in-bytes-if-the-memory-is-not-restricte seems to indicate the value is some very high number. Or is this handled by the JVMs cgroup layer?

Now I'm not 100% sure. I think it was a source code comment somewhere where I saw this special 0 treatment. I will have to refresh my memory...

Thank you for taking a look!

gerard-ziemski commented:
> I'll have a look tomorrow. In general unlimited means `-1` not `0`. cg v1 doesn't explicitly set it to a specified value if it is unlimited. It's a very large number. For cg v2 it's usually the string `max` as specified by the kernel docs of interface files.

Thank you for taking a look. I think I saw this special treatment of 0 value somewhere in source code. I will have to look.

(I wish I posted this PR sooner when all this was still fresh in my memory, but by the same token, I'm happy that at least we are taking a look at this right now and not a year from now...)

iklam commented:
> > I'll have a look tomorrow. In general unlimited means `-1` not `0`. cg v1 doesn't explicitly set it to a specified value if it is unlimited. It's a very large number. For cg v2 it's usually the string `max` as specified by the kernel docs of interface files.
> 
> Thank you for taking a look. 

According to
https://git.kernel.org/pub/scm/linux/kernel/git/tj/cgroup.git/tree/Documentation/admin-guide/cgroup-v2.rst

```
 memory.swap.max
	A read-write single value file which exists on non-root
	cgroups.  The default is "max".

	Swap usage hard limit.  If a cgroup's swap usage reaches this
	limit, anonymous memory of the cgroup will not be swapped out.
```

E.g.,

```
$ cat /sys/fs/cgroup/user.slice/memory.swap.max
max
```

I don't think there's a file under /sys/fs/cgroup where the numberic value of "0" means "unlimited". Instead, "0" means zero or "not allowed at all".

Otherwise, how can you express "don't allow any swapping for this container"?

I think the proper fix is:

```
void OSContainer::print_container_helper(outputStream* st, jlong j, const char* metrics) {
  st->print("%s: ", metrics);
- if (j > 0) {
+ if (j >= 0) {
```

> I think I saw this special treatment of 0 value somewhere in source code. I will have to look.

It's worthwhile scanning the source code to for such cases and possibly fixing them.

In most cases, I see that -1 means unlimited, and -2 (OSCONTAINER_ERROR) means error: 

```
jlong CgroupSubsystem::limit_from_str(char* limit_str) {
  if (limit_str == nullptr) {
    return OSCONTAINER_ERROR;
  }
  // Unlimited memory in cgroups is the literal string 'max' for
  // some controllers, for example the pids controller.
  if (strcmp("max", limit_str) == 0) {
    os::free(limit_str);
    return (jlong)-1;
  }
  julong limit;
  if (sscanf(limit_str, JULONG_FORMAT, &limit) != 1) {
    os::free(limit_str);
    return OSCONTAINER_ERROR;
  }
  os::free(limit_str);
  return (jlong)limit;
}
```

jerboaa commented on src/hotspot/os/linux/osContainer_linux.cpp:
```diff
@@ -137,9 +137,9 @@ jlong OSContainer::pids_current() {
-void OSContainer::print_container_helper(outputStream* st, jlong j, const char* metrics) {
+void OSContainer::print_container_helper(outputStream* st, jlong j, const char* metrics, boolean limit) {
   st->print("%s: ", metrics);
-  if (j > 0) {
```
I agree with @iklam. The fix for this seems to be to change the condition here to `j >= 0`.

jerboaa reviewed (changes requested):
The bug is a bit thin on the details. Can that be improved? I'm not aware that any of the APIs in hotspot return 0 for `unlimited`. If it is, it's a bug. It would be good to have this captured in a test somewhere. Would a test with `--memory 200M --memory-swap=200M` (no swap), trigger the said case? I think it would. For cg v1 the test case would probably also need a swappiness value of `0`.

gerard-ziemski commented:
> > > I'll have a look tomorrow. In general unlimited means `-1` not `0`. cg v1 doesn't explicitly set it to a specified value if it is unlimited. It's a very large number. For cg v2 it's usually the string `max` as specified by the kernel docs of interface files.
> > 
> > 
> > Thank you for taking a look.
> 
> According to https://git.kernel.org/pub/scm/linux/kernel/git/tj/cgroup.git/tree/Documentation/admin-guide/cgroup-v2.rst
> 
> ```
>  memory.swap.max
> 	A read-write single value file which exists on non-root
> 	cgroups.  The default is "max".
> 
> 	Swap usage hard limit.  If a cgroup's swap usage reaches this
> 	limit, anonymous memory of the cgroup will not be swapped out.
> ```
> 
> E.g.,
> 
> ```
> $ cat /sys/fs/cgroup/user.slice/memory.swap.max
> max
> ```
> 
> I don't think there's a file under /sys/fs/cgroup where the numberic value of "0" means "unlimited". Instead, "0" means zero or "not allowed at all".
> 
> Otherwise, how can you express "don't allow any swapping for this container"?
> 
> I think the proper fix is:
> 
> ```
> void OSContainer::print_container_helper(outputStream* st, jlong j, const char* metrics) {
>   st->print("%s: ", metrics);
> - if (j > 0) {
> + if (j >= 0) {
> ```
> 
>  > I think I saw this special treatment of 0 value somewhere in source code. I will have to look.
> 
> It's worthwhile scanning the source code to for such cases and possibly fixing them.
> 
> In most cases, I see that -1 means unlimited, and -2 (OSCONTAINER_ERROR) means error:
> 
> ```
> jlong CgroupSubsystem::limit_from_str(char* limit_str) {
>   if (limit_str == nullptr) {
>     return OSCONTAINER_ERROR;
>   }
>   // Unlimited memory in cgroups is the literal string 'max' for
>   // some controllers, for example the pids controller.
>   if (strcmp("max", limit_str) == 0) {
>     os::free(limit_str);
>     return (jlong)-1;
>   }
>   julong limit;
>   if (sscanf(limit_str, JULONG_FORMAT, &limit) != 1) {
>     os::free(limit_str);
>     return OSCONTAINER_ERROR;
>   }
>   os::free(limit_str);
>   return (jlong)limit;
> }
> ```

My original fix was just being extra careful and was fixing non-limit type values only.

However, looking at https://git.kernel.org/pub/scm/linux/kernel/git/tj/cgroup.git/tree/Documentation/admin-guide/cgroup-v2.rst I noticed a section on `Limits` which says:

`Limits are in the range [0, max] and defaults to "max", which is noop.
`

so 0 is a valid value, with no special meaning, so your suggestion on accepting 0 in all cases looks valid to me. Looks like the original implementation got this incorrect.

Thank you for the review and feedback!

gerard-ziemski commented:
> The bug is a bit thin on the details. Can that be improved? I'm not aware that any of the APIs in hotspot return 0 for `unlimited`. If it is, it's a bug. It would be good to have this captured in a test somewhere. Would a test with `--memory 200M --memory-swap=200M` (no swap), trigger the said case? I think it would. For cg v1 the test case would probably also need a swappiness value of `0`.

Actually hotspot is doing the opposite: it returns "unlimited" when it was supposed to return "0". You can see it right now if you run "jcmd PID VM.info" and look at "memory_swap_current_in_bytes". No special flags needed to see this.

Thank you for taking a look. I also agree that we should treat "0" as "0" for all cases.

gerard-ziemski commented:
> > > The bug is a bit thin on the details. Can that be improved? I'm not aware that any of the APIs in hotspot return 0 for `unlimited`. If it is, it's a bug. It would be good to have this captured in a test somewhere. Would a test with `--memory 200M --memory-swap=200M` (no swap), trigger the said case? I think it would. For cg v1 the test case would probably also need a swappiness value of `0`.
> 
> @gerard-ziemski Please add a test case for this in `test/hotspot/jtreg/containers/docker/TestMisc.java` akin to `testPrintContainerInfoActiveProcessorCount()`, setting the above memory limits and asserting that you get `memory_swap_current_in_bytes: 0` in the output. Since this is a cgv2 specific output you might need to add a condition on `Metrics.systemMetrics().getProvider().equals("cgroupv2")` and otherwise skip it.
> 
> > Actually hotspot is doing the opposite: it returns "unlimited" when it was supposed to return "0". You can see it right now if you run "jcmd PID VM.info" and look at "memory_swap_current_in_bytes". No special flags needed to see this.
> 
> Sorry for not being clear. This particular case exercises `CgroupV2Subsystem::print_version_specific_info`, which internally calls:
> 
> ```
>   char* mem_swp_current_str = mem_swp_current_val();
>   jlong swap_current = limit_from_str(mem_swp_current_str);
> ```
> 
> `swap_current` is then being passed to `OSContainer::print_container_helper`. So I meant the retrieval code which would return `-1` if it was `unlimited` (i.e. `swap_current == -1` if it was unlimited). In this case I'm pretty sure it returns `0`, but `OSContainer::print_container_helper` has the bug. That's what I meant with hotspot APIs. I should have been more precise.

Thank you for the feedback and the suggestion.

I chose to check that `unlimited` does not appear with `memory_swap_current_in_bytes` on the same line, not for `0` to exist, since it's possible that the current memory swap might be > 0, at some point later in the future.

jerboaa commented:
> I chose to check that `unlimited` does not appear with `memory_swap_current_in_bytes` on the same line, not for `0` to exist, since it's possible that the current memory swap might be > 0, at some point later in the future.

Why? If a container is being run with `--memory X` and --memory-swap X` then there must not be any swap used, so any value greater than `0` would be a bug and the assertion is safe. See https://docs.docker.com/config/containers/resource_constraints/#prevent-a-container-from-using-swap

I agree, though, that it might be better to move to a separate test. Up to you.

gerard-ziemski commented:
> > I chose to check that `unlimited` does not appear with `memory_swap_current_in_bytes` on the same line, not for `0` to exist, since it's possible that the current memory swap might be > 0, at some point later in the future.
> 
> Why? If a container is being run with `--memory X` and `--memory-swap X` (`X` being the same value) then there must not be any swap used, so any value greater than `0` would be a bug and the assertion is safe. See https://docs.docker.com/config/containers/resource_constraints/#prevent-a-container-from-using-swap

This bug fix is correcting the behavior where `0` should be displayed instead of `unlimited`, which is an approach that is flexible enough in case that this value happens to be greater than `0` in the future for any reason.

I'd really prefer to keep the test as simple as possible to cover this specific fix, which in my opinion it does.

Setting  `--memory X` and `--memory-swap X` introduces more variable and makes the test more complicated than it needs to be, in my opinion.

What do you think?

jerboaa commented:
> This bug fix is correcting the behavior where `0` should be displayed instead of `unlimited`. Strictly speaking that's all we need to test here. The test I proposed uses an approach that is flexible enough in case that this value happens to be greater than `0` in the future for any reason.

I disagree. What you are looking for is if the actual value is `0` you want `memory_swap_current_in_bytes` to report `0`. The test you have might end up running with any number `> 0` in the interface file, and your test would spuriously pass even with the actual bug present (since `unlimited` would not occur in this case). The only way to actually ensure you get `0` is by running the test in a container with swap turned off.

gerard-ziemski commented:
> > This bug fix is correcting the behavior where `0` should be displayed instead of `unlimited`. Strictly speaking that's all we need to test here. The test I proposed uses an approach that is flexible enough in case that this value happens to be greater than `0` in the future for any reason.
> 
> I disagree. What you are looking for is if the actual value is `0` you want `memory_swap_current_in_bytes` to report `0`. The test you have might end up running with any number `> 0` in the interface file, and your test would spuriously pass even with the actual bug present (since `unlimited` would not occur in this case). The only way to actually ensure you get `0` is by running the test in a container with swap turned off.

As long as the test gets run in an environment where we do get `0` at least some times, we are fine, but you are correct: to guarantee that we test the right thing, we would need to setup the container properly.

I feel like I don't have enough container knowledge to write such a test, without investing more time learning the basics. Would you kindly help and write the test you want please? At least the part that sets up the environment and starts the test (on Linux only?) with swap==0 ?

gerard-ziemski commented:
> > I feel like I don't have enough container knowledge to write such a test, without investing more time learning the basics. Would you kindly help and write the test you want please? At least the part that sets up the environment and starts the test (on Linux only?) with swap==0 ?
> 
> Sure. Here you go: [gerard-ziemski#1](https://github.com/gerard-ziemski/jdk/pull/1)

Thank you very much!

I will incorporate it into this fix shortly...

gerard-ziemski commented:
I have added your test with one change.

Instead of checking for strict syntax:

`out.shouldContain("memory_swap_current_in_bytes: 0"))`

we do:

```
            for (String s : str.split(System.lineSeparator())) {
                if (s.contains("memory_swap_current_in_bytes")) {
                    s.shouldContain("0");
                    s.shouldNotContain("unlimited");
                }
            }
```

I did this in anticipation of [JDK-8321932](https://bugs.openjdk.org/browse/JDK-8321932) which will change the layout.

> > > I feel like I don't have enough container knowledge to write such a test, without investing more time learning the basics. Would you kindly help and write the test you want please? At least the part that sets up the environment and starts the test (on Linux only?) with swap==0 ?
> > 
> > 
> > Sure. Here you go: [gerard-ziemski#1](https://github.com/gerard-ziemski/jdk/pull/1)
> 
> Thank you very much!
> 
> I will incorporate it into this fix shortly...

Ready for a re-review.

jerboaa reviewed (commented):
The test would be a bit easier to read if we extracted the loose matching to a function. Suggestion:

```
[...]
               out.shouldContain("memory_swap_max_limit_in_bytes");
               checkLineForValue(str, "memory_swap_max_limit_in_bytes", "0");
[...]

private static void checkLineForValue(String output, String match, String value) {
[...]
}
```
```

gerard-ziemski commented on test/hotspot/jtreg/containers/docker/TestContainerInfo.java:
```diff
@@ -0,0 +1,104 @@
+                    if (!s.contains("0")) {
+                        throw new RuntimeException("memory_swap_current_in_bytes NOT 0!");
+                    }
+                    if (s.contains("Unlimited")) {
```
done

gerard-ziemski commented on test/hotspot/jtreg/containers/docker/TestContainerInfo.java:
```diff
@@ -0,0 +1,104 @@
+            for (String s : str.split(System.lineSeparator())) {
+                if (s.contains("memory_swap_max_limit_in_bytes")) {
+                    if (!s.contains("0")) {
+                        throw new RuntimeException("memory_swap_max_limit_in_bytes NOT 0!");
```
done

tstuefe commented:
I have to beg out, too much to do before FOSDEM. But you have two reviewers already, so..

gerard-ziemski commented:
Thank you all for the reviews!

I will re-run Mach5 teir1-5 tests, just in case, before I push it...

gerard-ziemski commented:
I had to make a small fix in the test - I forgot to pass required argument.

Tested it with MACH5 tier1-5 and that looks fine.

jerboaa reviewed (approved):
Still fine.
